File: PracticeAlgorithm/ArrayChallenges.cs
using System;
using FluentAssertions;
using System.Collections.Generic;
namespace PracticeAlgorithm;

public class ArrayChallenges
{
    [Fact]
    public void MethodCaller()
    {
        var sum = SumOfEvenNumber(new int[] {2, 8, 1, 4, 3});
        sum.Should().Be(14);
    }

    [Fact]
    public void CheckSortNumber()
    {
        var result = SortNumber(new List<int>() {5, 8, 4, 1});
        result.Should().BeEquivalentTo(new List<int>() {1, 4, 5, 8});
    }

    [Fact]
    public void CheckSortString()
    {
        var result = SortString(new List<string>() {"abc", "bca", "aba"});
        result.Should().BeEquivalentTo(new List<string>() {"aba", "abc", "bca"});
    }

    [Fact]
    public void CheckArrayRotation()
    {
        var nums = new int[] {1, 2, 3, 4, 5, 6};
        var k = 3;
        RotateArrayToRight(nums, k);
        nums.Should().BeEquivalentTo(new int[] {4, 5, 6, 1, 2, 3});
    }

    [Fact]
    public void EvaluateReversePolishNotation()
    {
        var arr = new string[] {"2", "1", "+", "3", "*"};
        var result = ReversePolish(arr);
        result.Should().Be(9);
        var arr2 = new string[] {"4", "13", "5", "/", "+"};
        var result2 = ReversePolish(arr2);
        result2.Should().Be(6);
    }

    [Fact]
    public void CheckArrayRotation2()
    {
        var list = new List<int>() {1, 2, 3, 4, 5, 6, 7};
        RotateArray(list, 3);
        list.Should().BeEquivalentTo(new List<int>() {7, 6, 5, 1, 2, 3, 4});
    }

    [Theory]
    [MemberData(nameof(GetTestData))]
    public void Merge_ShouldBe_RemoveAndOrderData(List<int> l1, List<int> l2, List<int> list)
    {
        var result = MergeListUsingSet(l1, l2);
        result.Should().BeEquivalentTo(list);
    }

    private IEnumerable<int> MergeListUsingSet(IEnumerable<int> l1, IEnumerable<int> l2)
    {
        var set = new HashSet<int>();
        set.UnionWith(l1);
        set.UnionWith(l2);
        return set.OrderBy(x => x).ToList();
    }

    public static IEnumerable<object[]> GetTestData()
    {
        yield return new object[]
        {
            new List<int> {2, 8, 1, 8}, new List<int> {2, 9, 0, 10}, new List<int> {0, 1, 2, 8, 9, 10},
        };
    }
    [Theory]
    [InlineData(new int[]{1,2,3,0,0,0}, 3, new int[]{2,5,6}, 3)]
    public void Merge(int[] nums1, int m, int[] nums2, int n) {
        var merged = new int[m+n];
        var newNum1s= new int [m];
        var newNum2s= new int [n];
        for(int i =0; i<m; i++)
        {
            newNum1s[i]= nums1[i];
        }
        for(int i =0; i<n; i++)
        {
            newNum2s[i]= nums2[i];
        }
        
        merged= newNum1s.Concat(newNum2s).ToArray();
        Array.Sort(merged);
    }
    [Theory]
    [InlineData(new int[]{3,2,2,3}, 3)]
    public int RemoveElement(int[] nums, int val) {
        int index =0;
        for(int i =0; i<nums.Length; i++)
        {
            if(nums[i]!=val)
            {
                nums[index]=nums[i];
                index++;
            }
        }
        return index;
    }
    [Theory]
    [InlineData(new int[]{3,2,2,3})]
    public int RemoveDuplicates(int[] nums) {
        List<int> expectedNums = new List<int>();
        foreach(var num in nums)
        {
            if(!expectedNums.Contains(num))
            {
                expectedNums.Add(num) ;
            }
        }
        nums = expectedNums.ToArray();
        return nums.Length;
     
    }
    private static int SumOfEvenNumber(int[] nums)
    {
        return nums.Where(num => num % 2 == 0).Sum();
    }

    private List<int> SortNumber(List<int> nums)
    {
        for (int i = 0; i < nums.Count; i++)
        {
            for (int j = 0; j < nums.Count; j++)
            {
                if (nums[i] < nums[j])
                {
                    (nums[i], nums[j]) = (nums[j], nums[i]);
                }
            }
        }

        return nums;
    }

    private IEnumerable<string> SortString(IList<string> strgs)
    {
        for (var i = 0; i < strgs.Count; i++)
        {
            for (var j = 0; j < strgs.Count; j++)
            {
                if (string.CompareOrdinal(strgs[i], strgs[j]) < 0)
                {
                    (strgs[i], strgs[j]) = (strgs[j], strgs[i]);
                }
            }
        }

        return strgs;
    }

    private void RotateArrayToRight(int[] nums, int k)
    {
        if (nums.Count() < k)
        {
            k %= nums.Count();
        }
        //divide array into two parts 
        //rotate the first part
        //rotate the second part
        //rotate the whole array

        int a = nums.Length - k;
        Reverse(nums, 0, a - 1);
        Reverse(nums, a, nums.Length - 1);
        Reverse(nums, 0, nums.Length - 1);
    }

    private void Reverse(int[] nums, int left, int right)
    {
        while (left < right)
        {
            (nums[left], nums[right]) = (nums[right], nums[left]);
            left++;
            right--;
        }
    }
    [Theory]
    [MemberData(nameof(GetRotateTestData))]
    public void Rotate(int[][] matrix)
    {
        for (int i = 0; i < matrix.Length; i++)
        {
            for (int j = i + 1; j < matrix.Length; j++)
            {
                (matrix[i][j], matrix[j][i]) = (matrix[j][i], matrix[i][j]);
            }
        }

        foreach (var row in matrix)
        {
            Array.Reverse(row);
        }
        
    }
    public static IEnumerable<object[]> GetRotateTestData()
    {
        yield return new object[] 
        { 
            new int[][] 
            {
                new int[] { 1, 2, 3 },
                new int[] { 4, 5, 6 },
                new int[] { 7, 8, 9 }
            }
        };
    }
    ///  Naive Approach
    ///that this problem can be solved by using a stack.We can loop through each element
    //in the given array.When it is a number, push it to the stack.When it is an operator,
    //pop two numbers from the stack, do the calculation, and push back the result.
    private int ReversePolish(string[] arr)
    {
        var stack = new Stack<int>();
        foreach (var item in arr)
        {
            if (item == "+" || item == "/" || item == "*" || item == "-")
            {
                switch (item)
                {
                    case "+":
                    {
                        var num1 = stack.Pop();
                        var num2 = stack.Pop();
                        stack.Push(num1 + num2);
                    }
                        break;
                    case "/":
                    {
                        var num1 = stack.Pop();
                        var num2 = stack.Pop();
                        stack.Push(num2 / num1);
                    }

                        break;
                    case "*":
                    {
                        var num1 = stack.Pop();
                        var num2 = stack.Pop();
                        stack.Push(num1 * num2);
                    }
                        break;
                    case "-":
                    {
                        var num1 = stack.Pop();
                        var num2 = stack.Pop();
                        stack.Push(num2 - num1);
                    }
                        break;
                }
            }
            else
            {
                stack.Push(Convert.ToInt32(item));
            }
        }

        return stack.Pop();
    }

    [Fact]
    public void BeepVowelsShouldReturnVowelsReplacedByAstrix()
    {
        BeepVowels("skateboard").Should().Be("sk*t*b**rd");
        BeepVowels("slipper").Should().Be("sl*pp*r");
        BeepVowels("brisk morning").Should().Be("br*sk m*rn*ng");
    }

    private string BeepVowels(string str)
    {
        var strToLower = str.ToLower().ToCharArray();
        string newString = "";
        char[] vowels = new char[] {'a', 'e', 'i', 'o', 'u'};
        foreach (char c in strToLower)
        {
            int index = Array.IndexOf(vowels, c);
            if (index > -1)
            {
                newString += '*'.ToString();
            }
            else
            {
                newString += c.ToString();
            }
        }

        return newString;
    }

    [Fact]
    public void CreateAcronomyShouldBeReturnAcronomyOfTheSentence()
    {
        CreateAcronomy("New York").Should().Be("NY");
        CreateAcronomy("same stuff different day").Should().Be("SSDD");
        CreateAcronomy("Laugh out loud").Should().Be("LOL");
        CreateAcronomy("don't over think stuff").Should().Be("DOTS");
    }

    private string CreateAcronomy(string str)
    {
        string acronomy = "";
        foreach (string c in str.ToUpper().Split(" "))
        {
            acronomy += c.Substring(0, 1);
        }

        return acronomy;
    }

    [Fact]
    public void TestName()
    {
        WordCount("open the window please", new string[] {"please", "open", "sorry"}).Should().Be(2);
        WordCount("drive to the cinema", new string[] {"the", "driver"}).Should().Be(1);
        WordCount("can I have that can", new string[] {"can", "I"}).Should().Be(3);
    }

    private int WordCount(string str, string[] arr)
    {
        var words = str.Split(" ");
        int count = 0;
        foreach (string x in words)
        {
            int index = Array.IndexOf(arr, x);
            if (index > -1) count++;
        }

        return count;
    }

    private void RotateArray(IList<int> arr, int k)
    {
        if (k > arr.Count)
            k = k % arr.Count;
        for (var i = 0; i < k; i++)
        {
            for (var j = arr.Count - 1; j > 0; j--)
            {
                (arr[j - 1], arr[j]) = (arr[j], arr[j - 1]);
            }
        }
    }

    public IList<IList<int>> PermuteUnique(int[] nums)
    {
        List<IList<int>> permution = new List<IList<int>>();

        return permution;
    }

    private void BackTracking(List<List<int>> res, int[] nums, Dictionary<int, bool> used)
    {
    }

    [Fact]
    public void AlternatingCapsShouldBeAlteringTheSentenceStartingByLoweringTheWordIfNumberIsEven()
    {
        AlternatingCaps("take them to school").Should().Be("take THEM to SCHOOL");
        AlternatingCaps("What did ThEy EAT before?").Should().Be("what DID they EAT before?");
    }

    private string AlternatingCaps(string str)
    {
        var newArr = str.Split(" ");
        var result = new List<string>();
        for (int i = 0; i < newArr.Length; i++)
        {
            if (i % 2 == 0)
            {
                result.Add(newArr[i].ToLower());
            }
            else
            {
                result.Add(newArr[i].ToUpper());
            }
        }

        return string.Join(" ", result);
    }

    [Fact]
    public void CommonElementsShouldReturnCommonElementsOfAnyTypes()
    {
        string[] arr1 = new string[] {"a", "c", "d", "b"};
        string[] arr2 = new string[] {"b", "a", "y"};
        CommonElements(arr1, arr2).Should().BeEquivalentTo(new string[] {"a", "b"});
        int[] arr3 = new int[] {4, 7};
        int[] arr4 = new int[] {32, 7, 1, 4};
        CommonElements(arr3, arr4).Should().BeEquivalentTo(new int[] {4, 7});
    }

    private T[] CommonElements<T>(T[] arr1, T[] arr2)
    {
        var result = new List<T>();
        foreach (var arr in arr1)
        {
            var foundInResult = result.FirstOrDefault(x => EqualityComparer<T>.Default.Equals(x, arr));
            var foundInArr2 = arr2.FirstOrDefault(x => EqualityComparer<T>.Default.Equals(x, arr));
            if ((foundInResult == null || EqualityComparer<T>.Default.Equals(foundInResult, default(T))) &&
                (foundInArr2 != null && !EqualityComparer<T>.Default.Equals(foundInArr2, default(T))))
            {
                result.Add(arr);
            }
        }

        return result.ToArray();
    }

    [Theory]
    [InlineData(new int[] {1, 2, 3, 4}, 2)]
    [InlineData(new int[] {1, 2, 3, 4, 5}, 2)]
    public void ArrayChunk(int[] arr, int size)
    {
        //   var newArr = arr.ToList();
        //   var twoDList =new List<List<int>>();
        //   int startIndex = 0;
        //   while(startIndex<newArr.Count)
        //   {
        //   var destArray = newArr.ToList().GetRange(startIndex, Math.Min(size, newArr.Count-startIndex));
        //   twoDList.Add(destArray);
        //   startIndex +=size;
        //   }
        //   var lastArray = twoDList.ToArray();
        //Or

        int startIndex = 0;
        var result = new List<List<int>>();
        while (startIndex < arr.Length)
        {
            int newSize = Math.Min(size, arr.Length - startIndex);
            var destination = new int [newSize];
            Array.Copy(arr, startIndex, destination, 0, newSize);
            result.Add(destination.ToList());
            startIndex += newSize;
        }
    }
        [Theory]
        [InlineData("abcdefg",3,new string[]{"abc","de", "fg"})]
        [InlineData("abcdefg",6, new string[]{"ab","c","d","e","f", "g"})]
    public void StringChunks(string s, int chunk, string[] result)
    {
        var length = s.Length;
        var chunkSize = length / chunk;
        var remainder = length % chunk;
        var startIndex = 0;
        var newStr = new string[chunk];
        for (var i = 0; i < chunk; i++)
        {
            var chuckSizeLength = chunkSize;
            if (i < remainder)
            {
                chuckSizeLength++;
            }
            newStr[i] = s.Substring(startIndex, chuckSizeLength);
            startIndex += chuckSizeLength;
        }

        newStr.Should().BeEquivalentTo(result);

    }
    
    [Theory]
    [InlineData(new int[]{7,1,5,3,6,4}, 7)]
    public void CalculateMaxProfitShouldBeEqualToResult(int[] prices, int result) {
        var b=0;
        var profit =0;
        for(var s =1; s<prices.Length; s++ )
        {
            var currentProfit = prices[s] - prices[b];
            if(currentProfit > 0)
            {
                profit += currentProfit;
            } 
            b++;
        }
        profit.Should().Be(result);
    }
    [Theory]
    [InlineData(new int[]{1,2,3,4,5,6,7}, 3)]
    public void RotateToRightByK(int[] nums, int k) {
        //find reminder of k divided by nums.length because k
        //is more than nums.length we don't need to iterate huge iterations.
        // first of all reverse entire the array.
        // reverse the array from 0th index to k-1 index.
        // reverse the array from kth index to array.length-1 index.
        k%=nums.Length;
        ReverseNums(nums, 0, nums.Length-1);
        ReverseNums(nums, 0, k-1);
        ReverseNums(nums, k, nums.Length-1);
    }
    private void ReverseNums(int[] nums, int l, int r)
    {
        while(l<=r)
        {
            var temp = nums[l];
            nums[l]= nums[r];
            nums[r]=temp;
            l++;
            r--;
        }
    }
}

File: JavascriptChallenges/arrayChallenges.js
// Write a function `stayPositive` that accepts an array of numbers as an argument. The function should
// return an array containing only the positive numbers.
// const stayPositive = (arr) =>
// {
//     let positive =[];
//     for(let x of arr)
//     {
//         if(x>0)
//         {
//         positive.push(x)
//         }
//     }
//     return positive;
// }
// console.log(stayPositive([10, -4, 3, 6])); // [10, 3, 6]
// console.log(stayPositive([-5, 11, -40, 30.3, -2])); // [11, 30.3]
// console.log(stayPositive([-11, -30])); // []

// Write a function `bleepVowels` that accepts a string as an argument. The function should return
// a new string where all vowels are replaced with `*`s. Vowels are the letters a, e, i, o, u.
//  const bleepVowels = (str) => 
//  {
//     const strToLower = str.toLowerCase();
//      let newString ="";
//      let vowels = ["a", "e", "i", "o", "u"];

//     for(let c of strToLower)
//     {
//         if(vowels.indexOf(c) >-1)
//         {
//             newString+='*' 
//         }
//         else
//         {
//             newString+=c;  
//         }
//         // switch(c)
//         // {
//         //     case 'a':
//         //     case 'e': 
//         //     case 'i':
//         //     case 'o':
//         //     case 'u':  
//         //     newString+='*' 
//         //     break;
//         //     default:
//         //         newString+=c;  
//         // }
//     }
//     return newString;
//  }
// console.log(bleepVowels("skateboard")); // 'sk*t*b**rd'
// console.log(bleepVowels("slipper")); // 'sl*pp*r'
// console.log(bleepVowels("range")); // 'r*ng*'
// console.log(bleepVowels("brisk morning")); // 'br*sk m*rn*ng'

// Write a function `filterLongWords` that accepts an array of strings as an argument. The function
// should return a new array containing only the strings that are less than 5 characters long.
// const filterLongWords =(arr) =>
// {
//     const newArr= []
//     for(let a of arr)
//     {
//         if(a.length <5)
//         {
//             newArr.push(a)
//         }
//     }
//     return newArr;
// }
// console.log(filterLongWords(["kale", "cat", "retro", "axe", "heirloom"]));
// // ['kale', 'cat', 'axe']

// console.log(filterLongWords(["disrupt", "pour", "trade", "pic"]));
// // ['pour', 'pic']

// Write a function `numOdds` that accepts an array of numbers as an argument. The function should
// return a number representing the count of odd elements in the array.
//   const numOdds = (arr) =>
//   {
//     let count =0;
//     for(let n of arr)
//     {
//         if(n%2===1)
//         {
//           count++  
//         }
//     }
//     return count;
//   }
// console.log(numOdds([4, 7, 2, 5, 9])); // 3
// console.log(numOdds([11, 31, 58, 99, 21, 60])); // 4
// console.log(numOdds([100, 40, 4])); // 0

// Write a function `stringsToLengths` that accepts an array of strings as an argument. The function
// should return a new array containing the lengths of the elements of the original array.
// const stringsToLengths = (arr) => 
// {
//     const result =[]
//     for(let a of arr)
//     {
//        result.push(a.length)
//     }
//     return result;
// }
// console.log(stringsToLengths(["belly", "echo", "irony", "pickled"]));
// // [5, 4, 5, 7]

// console.log(stringsToLengths(["on", "off", "handmade"]));
// // [2, 3, 8]
// Write a function `divisors` that accepts a number as an argument. The function should return an
// array containing all positive numbers that can divide into the argument.
// const divisors = (n) => 
// {
//     const result =[];
//     for(let i=1; i<=n; i++)
//     {
//         if(n%i===0)
//         {
//             result.push(i)
//         }
//     }
//     return result;
// }
// console.log(divisors(15)); // [1, 3, 5, 15]
// console.log(divisors(7)); // [1, 7]
// console.log(divisors(24)); // [1, 2, 3, 4, 6, 8, 12, 24]

// Write a function `makeAcronym` that accepts a sentence string as an argument. The function should
// return a string containing the first character of each word in the sentence.
//  const makeAcronym = (str) => {
//     let acronomy ="";
//     for(let c of str.toUpperCase().split(" "))
//     {
//         acronomy += c.slice(0,1);
//     }
//   return acronomy;
//  }
// console.log(makeAcronym("New York")); // NY
// console.log(makeAcronym("same stuff different day")); // SSDD
// console.log(makeAcronym("Laugh out loud")); // LOL
// console.log(makeAcronym("don't over think stuff")); // DOTS

// Write a function `wordCount(sentence, targetWords)` that accepts a sentence string and an array of
// `targetWords`. The function should return a count of the number of words of the sentence that are
// in `targetWords`.
//  const wordCount =(str, arr) =>
//  {
//     const words = str.split(" ");
//     let count =0;
//     for(let x of words)
//     {
//         let index = arr.indexOf(x);
//         if(index >-1)count++;
//     }
//     return count;
//  }
// console.log(wordCount("open the window please", ["please", "open", "sorry"])); // 2
// console.log(wordCount("drive to the cinema", ["the", "driver"])); // 1
// console.log(wordCount("can I have that can", ["can", "I"])); // 3

// Write a function `yourAverageFunction` that accepts an array of numbers as an argument. The
// function should return the average of all elements of the array. If the input array is empty,
// then the function should return null.
// const yourAverageFunction = (nums) =>
// {
//     if(nums.length==0) return null;
//     let sum =0;
//      for(let n of nums)
//      {
//         sum+=n;
//      }
//      return sum/nums.length;
// }
// console.log(yourAverageFunction([5, 2, 7, 24])); // 9.5
// console.log(yourAverageFunction([100, 6])); // 53
// console.log(yourAverageFunction([31, 32, 40, 12, 33])); // 29.6
// console.log(yourAverageFunction([])); // null

// Write a function `chooseDivisibles(numbers, target)` that accepts an array of numbers and a
// target number as arguments. The function should return an array containing elements of the original
// array that are divisible by the target.
// const chooseDivisibles =(nums, target)=>
// {
//      const result =[];
//      for(let num of nums)
//      {
//         if(num%target===0)
//         {
//             result.push(num)
//         }
//      }
//    return result;
// }
// console.log(chooseDivisibles([40, 7, 22, 20, 24], 4)); // [40, 20, 24]
// console.log(chooseDivisibles([9, 33, 8, 17], 3)); // [9, 33]
// console.log(chooseDivisibles([4, 25, 1000], 10)); // [1000]


// Write a function `maximum` that accepts an array of numbers as an argument. The function should 
// return the largest number of the array. If the array is empty, then the function should return null.
// const maximum = (nums)=> 
// {
//     if(nums.length==0) return null;
//     let largest = nums[0];
//     for(let n =1; n<nums.length; n++)
//     {
//         if(nums[n]>largest)
//         {
//             largest = nums[n];
//         }

//     }
// return largest;
// }
// console.log(maximum([5, 6, 3, 7])); // 7
// console.log(maximum([17, 15, 19, 11, 2])); // 19
// console.log(maximum([])); // null

// Write a function `lengthiestWord` that accepts a sentence string as an argument. The function should
// return the longest word of the sentence. If there is a tie, return the word that appears later 
// in the sentence.
// const lengthiestWord = (str) => 
// {
//     let longestWord = "";
//     const words = str.split(" ");
//     for(let word of words)
//     {
//         if(word.length > longestWord.length)
//         {
//             longestWord = word;
//         }
//     }
//     return longestWord;
// }
// console.log(lengthiestWord("I am pretty hungry")); // 'pretty'
// console.log(lengthiestWord("we should think outside of the box")); // 'outside'
// console.log(lengthiestWord("down the rabbit hole")); // 'rabbit'
// console.log(lengthiestWord("simmer down")); // 'simmer'

// Write a function `alternatingCaps` that accepts a sentence string as an argument. The function should
// return the sentence where words alternate between lowercase and uppercase.
// const alternatingCaps =(str) => 
// {
//     const words = str.split(" ");
//     let isLower = true;
//     let result =[];
//     for(let word of words)
//     {
//         if(isLower)
//         {
//             isLower = false;
//            result.push(word.toLowerCase())
//         } 
//         else
//         {
//             result.push(word.toUpperCase());
//             isLower=true;
//         }
//     }
//     return result.join(" ");
// }
// console.log(alternatingCaps("take them to school")); // 'take THEM to SCHOOL'
// console.log(alternatingCaps("What did ThEy EAT before?")); // 'what DID they EAT before?'

// Write a function `numberRange(min, max, step)` that accepts three numbers as arguments, `min`,
// `max`, and `step`. The function should return all numbers between `min` and `max` at `step` intervals.
// `min` and `max` are inclusive.
// const numberRange = (min, max, step)=>
// {
//     const result = [];
//     for(let n=min; n<=max; n+=step)
//     {
//             result.push(n);
//     }
//     return result;
// }

// console.log(numberRange(10, 40, 5)); // [10, 15, 20, 25, 30, 35, 40]
// console.log(numberRange(14, 24, 3)); // [14, 17, 20, 23]
// console.log(numberRange(8, 35, 6)); // [8, 14, 20, 26, 32]

// Write a function `removeShortWords` that accepts a sentence string as an argument. The function
// should return a new sentence where all of the words shorter than 4 characters are removed.
// const removeShortWords = (str) => 
// {
//     const words = str.split(" ");
//     const result = [];
//     for(let word of words )
//     {
//         if(word.length >=4)
//         {
//             result.push(word)
//         }
//     }
//    return result.join(" ");
// }
// console.log(removeShortWords("knock on the door will you")); // 'knock door will'
// console.log(removeShortWords("a terrible plan")); // 'terrible plan'
// console.log(removeShortWords("run faster that way")); // 'faster that'


// Write a function `commonElements` that accepts two arrays as arguments. The function should return
// a new array containing the elements that are found in both of the input arrays. The order of
// the elements in the output array doesn't matter as long as the function returns the correct elements.
const commonElements = (arr1, arr2) =>
{
    const result =[];
    for(let arr of arr1)
    {
        if(!result.includes(arr) && arr2.includes(arr))
        {
            result.push(arr)
        }
    }
    return result;
}
let arr1 = ["a", "c", "d", "b"];
let arr2 = ["b", "a", "y"];
console.log(commonElements(arr1, arr2)); // ['a', 'b']

let arr3 = [4, 7];
let arr4 = [32, 7, 1, 4];
console.log(commonElements(arr3, arr4)); // [4, 7]